src/hotspot/share/gc/shared/memAllocator.cpp

```diff
@@ -384,12 +384,18 @@ void MemAllocator::mem_clear(HeapWord* mem) const {
 
 oop MemAllocator::finish(HeapWord* mem) const {
   assert(mem != NULL, "NULL object pointer");
+  if (UseBiasedLocking) {
+    oopDesc::set_mark(mem, _klass->prototype_header());
```

Comment:
Not enough, I think, because there is a `oopDesc::set_mark` in the block below that would overwrite it.

Reply:
I think this one would be cleaner:

```
if (UseCompactObjectHeaders) {
  oopDesc::release_set_mark(mem, _klass->prototype_header());
} else {
  if (UseBiasedLocking) {
    oopDesc::set_mark(mem, _klass->prototype_header());
  } else {
    // May be bootstrapping
    oopDesc::set_mark(mem, markWord::prototype());
  }
  // Need a release store to ensure array/class length, mark word, and
  // object zeroing are visible before setting the klass non-NULL, for
  // concurrent collectors.
  if (!UseCompactObjectHeaders) {
    oopDesc::release_set_klass(mem, _klass);
  }
}
```

Reply:
Copy this to `HeapShared::copy_roots` too.